ghcide/session-loader/Development/IDE/Session.hs

```diff
@@ -798,6 +800,63 @@ setNameCache :: IORef NameCache -> HscEnv -> HscEnv
 #endif
 setNameCache nc hsc = hsc { hsc_NC = nc }
 
+-- This function checks then important property that if both p and q are home units
+-- then any dependency of p, which transitively depends on q is also a home unit.
+-- GHC had an implementation of this function, but it was horribly inefficient
+-- We should move back to the GHC implementation on compilers where
+-- https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12162 is included
+checkHomeUnitsClosed' ::  UnitEnv -> OS.Set UnitId -> [DriverMessages]
+checkHomeUnitsClosed' ue home_id_set
```

Comment:
The process is to construct a graph of all home units and their immediate dependencies, find all transitive dependency edges and then inverse that graph and check what is reachable from all home units. What remains must be exactly all home units.

Is that correct? Only the immediate dependencies of each home unit? Or does `UnitInfoMap` store all transitive dependencies?

There is the assumption, that there is no external unit that depends on home units but isn't dependent on by another home unit. In other words, all root nodes are home units. I think that's a fair assumption.

One question, wouldn't be more efficient to transpose the graph first, compute the reachable set, invert that and then perform the downsweep? I don't immediately see how performing the downsweep right away is getting rid of any edges.

Reply:
Thanks for the clarification!